lib/src/query.dart

```diff
@@ -0,0 +1,766 @@
+    final listPtr = Pointer<Int32>.allocate(count: length);
+    try {
+      for (int i=0; i<length; i++) {
+        listPtr.elementAt(i).store(_list[i]);
+      }
+      return func(qbPtr, propertyId, listPtr, length);
+    }finally {
+      listPtr.free();
+    }
+  }
+
+  // TODO replace duplication with implementation above, when fix is in
+  int _opList64(Pointer<Void> qbPtr, QueryCondition qc, obx_qb_cond_operator_in_dart_t<Int64> func) {
+    int propertyId = qc._propertyId;
+    int length = _list.length;
+    final listPtr = Pointer<Int64>.allocate(count: length);
+    try {
+      for (int i=0; i<length; i++) {
+        listPtr.elementAt(i).store(_list[i]);
+      }
+      return func(qbPtr, propertyId, listPtr, length);
+    }finally {
+      listPtr.free();
+    }
+  }
+
+}
+
+class DoubleCondition extends Condition<double> {
+  DoubleCondition(ConditionOp op, ConditionType type, double value, double value2)
+      : super(op, type, value, value2);
+
+  int _op1(Pointer<Void> qbPtr, QueryCondition qc, obx_qb_cond_operator_1_dart_t<double> func) {
+    return func(qbPtr, qc._propertyId, _value);
+  }
+}
+
+/**
+ * The first element of the chain
+ * contains the enum representation of
+ * the to-be-constructed query builder.
+ * This design allows nested chains inside
+ * the chain.
+ */
+class QueryCondition {
+  bool _root = true;
+  int _entityId, _propertyId;
+  Condition _condition;
+  List<List<QueryCondition>> _dnf; // all
+  int _group = 1;
+
+  QueryCondition(this._entityId, this._propertyId, this._condition);
+
+  // && is not overridable
+  QueryCondition operator&(QueryCondition rh) => and(rh);
+
+  // || is not overridable
+  QueryCondition operator|(QueryCondition rh) => or(rh);
+
+  // TODO remove later
+  void debug() {
+    _dnf?.forEach((qc) => qc.map((c) => c._condition).forEach((c) => print("${c._value}")));
+    print("dnf size: ${_dnf.length}");
+  }
+
+  void _initDnfList(QueryCondition qc) {
+    _dnf ??= <List<QueryCondition>>[];
+    if (_dnf.length < _group) {
+      _dnf.add(<QueryCondition>[]);
+    }
+    _dnf[_group - 1].add(qc);
+  }
+
+  QueryCondition or(QueryCondition rh) {
+    rh._root = false;
+    if (_dnf == null) {
+      _initDnfList(this);
+    }
+    _group++;
+    _initDnfList(rh);
+    return this;
+  }
+
+  QueryCondition and(QueryCondition rh) {
+    rh._root = false;
+    if (_dnf == null) {
+      _initDnfList(this);
+    }
+    _initDnfList(rh);
+    return this;
+  }
+}
+
+class Query<T> {
+  Pointer<Void> _query;
+  Box<T> _box;
+
+  // package private ctor
+  Query._(Box<T> box, Pointer<Void> qb) {
+    _box = box;
+    _query = checkObxPtr(bindings.obx_query_create(qb), "create query", true);
+  }
+
+  int count() {
+    final ptr = Pointer<Uint64>.allocate(count: 1);
+    try {
+      checkObx(bindings.obx_query_count(_query, ptr));
+      return ptr.load();
+    }finally {
+      ptr.free();
+    }
+  }
+
+  // TODO does dart have a dtor/finalizer?j
```

Comment:
unfortunately, it doesn't... seems there's no way to have this called automatically so you can safely remove the TODO

Reply:
I was thinking about [java's autocloseable](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html) lately. It would be interesting to emulate this somehow.

Reply:
yeah, that'd probably work, something like `with(arg1, arg2, arg3, fn(){})`...